esp-hal/src/gpio/mod.rs

```diff
@@ -961,80 +967,64 @@ where
     }
 }
 
-/// General Purpose Input/Output driver
-pub struct Io {
-    _io_mux: IO_MUX,
-    /// The pins available on this chip
-    pub pins: Pins,
-}
-
-impl Io {
-    /// Initialize the I/O driver.
-    pub fn new(gpio: GPIO, io_mux: IO_MUX) -> Self {
-        Self::new_with_priority(gpio, io_mux, crate::interrupt::Priority::min())
-    }
-
-    /// Initialize the I/O driver with a interrupt priority.
-    ///
-    /// This decides the priority for the interrupt when using async.
-    pub fn new_with_priority(
-        mut gpio: GPIO,
-        io_mux: IO_MUX,
-        prio: crate::interrupt::Priority,
-    ) -> Self {
-        gpio.bind_gpio_interrupt(gpio_interrupt_handler);
-        crate::interrupt::enable(crate::peripherals::Interrupt::GPIO, prio).unwrap();
-
-        Self::new_no_bind_interrupt(gpio, io_mux)
-    }
-
-    /// Initialize the I/O driver without enabling the GPIO interrupt or
-    /// binding an interrupt handler to it.
-    ///
-    /// *Note:* You probably don't want to use this, it is intended to be used
-    /// in very specific use cases. Async GPIO functionality will not work
-    /// when instantiating `Io` using this constructor.
-    pub fn new_no_bind_interrupt(gpio: GPIO, _io_mux: IO_MUX) -> Self {
-        Io {
-            _io_mux,
-            pins: gpio.pins(),
+impl<const GPIONUM: u8> InterruptConfigurable for GpioPin<GPIONUM>
+where
+    Self: GpioProperties,
+{
+    fn set_interrupt_handler(&mut self, handler: InterruptHandler) {
+        raise_gpio_interrupt_priority(handler.priority());
+        USER_INTERRUPT_HANDLERS[GPIONUM as usize].store(handler.handler());
+    }
+}
+
+static MAX_PRIO: AtomicUsize = AtomicUsize::new(Priority::None as usize);
+
+fn raise_gpio_interrupt_priority(priority: Priority) {
+    #[cold]
+    fn do_raise_priority_level(current_prio: usize, priority: Priority) {
+        let priority_level = priority as usize;
+        loop {
+            match MAX_PRIO.compare_exchange(
+                current_prio,
+                priority_level,
+                Ordering::Relaxed,
+                Ordering::Relaxed,
+            ) {
+                Ok(_) => {
+                    unsafe {
+                        crate::interrupt::bind_interrupt(
+                            crate::peripherals::Interrupt::GPIO,
+                            gpio_interrupt_handler,
+                        );
+                    }
+                    crate::interrupt::enable(crate::peripherals::Interrupt::GPIO, priority)
+                        .unwrap();
+                }
+                Err(p) if p < priority_level => {}
```

Comment:
`current_prio = p`? Otherwise this loops forever

Reply:
That's a very good point, thanks